### ISSUE TITLE
macro atoms calculations with very long run times

### ISSUE THREAD

**Issue by kslong** [sirocco-rt/sirocco#669]
In attempting to debug #659  I have been trying to compare macro atom and simple atom runs of optically thick stellar winds.  A number of the macro atom calculations require many (10x) as long as the analogous simple atom calculation, and some just seem to get permanently stuck.  Presumably all of this has to do with optical depths, or the time to execute a single macro atom interaction, but, if we are going to be working with optically thick models in macro atom mode, a solution is needed.

Here are different examples.  The first one runs reasonably and the remainder run long

[star_macro.pf.txt](https://github.com/agnwinds/python/files/4275906/star_macro.pf.txt)
[star_macro_pretty_thick.pf.txt](https://github.com/agnwinds/python/files/4275909/star_macro_pretty_thick.pf.txt)
[star_macro_thick.pf.txt](https://github.com/agnwinds/python/files/4275911/star_macro_thick.pf.txt)

**Comment by jhmatthews** [comment]
Just to summarise issue here -- this is probably due to the known issues with macro-atom performance in dense regions or where any processes are close to detailed balance (see e.g. #54). I think density/departure from LTE is the driver, rather than optical depth per se, but clearly they are related. One of the proposed solutions is  #423 , but getting this to work in the general code and not just the emissivity calculation would need some time investment as infrastructure is needed to store transition probability matrices.

**Comment by kslong** [comment]
@jhmatthews   I would appreciate your comments on this.  I reran the star_marcro_thick model today and with 8 threads on my MAC, the star_macro_thick_model took less time (2566 s) with jumps than with matrix (4768) on the dev branch. This was not what I was expecting.

**Comment by kslong** [comment]
The difference in the spectra produced using matrix and jumps modes look very similar, as do the electron and radiation temperature profiles, so there is no indication that either of the approaches is wrong.  The only question is why the time is different.

And just to be clear most of the time is being spent in photon transport of inization cycles.  Furthermore, these routines were run with the -p options so that the number of photons increase between cycles.  In ionization cycle 2, the photon transport took 797 s in the matrix method and 156 with jumps.  In cycle 30, with many more photons, it was 212 with matrix and 166 with jumps.  So most of the speed difference was in the early ionization cycles.  

Here are the .sig files

[z.4James.211111.tar.gz](https://github.com/agnwinds/python/files/7521915/z.4James.211111.tar.gz)

**Comment by jhmatthews** [comment]
I agree this is a little surprising. Do you have the .sig files you can post?

**Comment by kslong** [comment]
@saultyevil @smangham   I'd be interested in your take on this.  This was a .pf file of a star with a think wind that was running quite long at some time in the past.  Currently, it does run both in matrix and in jump mode.  The conventional wisdom is that matrix mode is faster but that does not seem to be happening.  I have just run the files in single processor mode (with the -p option) so I could see what was happening with gprof.  

On the computers I was running (there were two but they should be very similar in terms of speed, and I have gotten qualitatively the same result on my mac), but the jump version of the program runs in 51400s (according to the .sig file), while the matrix version runs in 92600s. 

The profile files, along with the .pf files and .sig files are in the attached file.  If you look at them, one sees that most of the processes are taking about the same amount of time, but the subroutine fraction takes much longer in the matrix version, despite being called fewer times.  Also if you look, the profiler gives much less time 22286 and 28707 for all the processes, much less than the wallclock time.  Is this expected?  Does it indicate anything?   Anyway, I'd appreciate your thoughts, as would @jhmatthews .

[z.today.211112.tar.gz](https://github.com/agnwinds/python/files/7530031/z.today.211112.tar.gz)

**Comment by kslong** [comment]
I have verified I get exactly the same result as above, when running on the same computer, and specifically that gprof from the matrix mode vs jump mode look quite different, with fraction requiring much more time in matirx mode than in jump mode, despite being called fewer times.

**Comment by kslong** [comment]
@jhmatthews    I have been using valgrind on this problem (which actually gives a clearer rsults than gprof.  I think it is pretty clear that the reason the matrix mode is taking longer than jump mode is due to what is happening in matom_emit.  It looks like the underlying problem is that emit_matom is spending a lot of time calling the numerical integration routines via alpha_sp.   Here is a figure indicationg where the time is being spent in matrix mode.  Somehow we need to avoid/store this integrations.

It looks like these are the spontaneous recombination rates, which depend on the temperature, but which could be calculated only once for each cell in each cycle


<img width="485" alt="screenshot_1011" src="https://user-images.githubusercontent.com/2530901/142257706-6821d596-db35-4f42-83d6-58fdcfeffa68.png">

**Comment by jhmatthews** [comment]
Aha. I think there's something easy that can be done here. In matom we do the following:

```
sp_rec_rate = mplasma->recomb_sp[config[uplvl].bfd_indx_first + n];
```

that is, we've already stored these rates for this exact reason. It looks like we didn't update emit_matom to use this. A good example of why to avoid code duplication, I suppose. I think the fix might be as simple as moving a line like this into emit_matom but we might need to check whether there's an energy weighting or anything. 

Certainly there's no good reason why the matrix method should have to generate more photons, so I was confused for a while - if it's what I've suggested, that clears it up.

**Comment by kslong** [comment]
Yes that change sped up the matrix method for this significantly.  With this fix, I found that for 8 threads and running with the -p option,  the matrix method ran in 2749 s and the jump method in 2566 s.  So jumps are still somewhat faster, but before this change the matrix method ran in 4768 s.  The spectra for the matrix method were exactly the same, before and after, so the actual radiative transfer was not affected.  I still would like to understand why the matrix method is not significantly faster; I will try running valgrind again on the modified code.

**Comment by kslong** [comment]
I've looked at this a bit more for the full run.  If you look at the current version of xsig for running with jumps vs matrix mode, what one observes is that matrix mode is much faster in early cycles but jumps catches up in late cycles.    Here is a plot.  Note that the general increase in times is due to the -p option

![times](https://user-images.githubusercontent.com/2530901/142299277-daab6ad4-8f87-4540-b756-cbbc6e3dbc92.png)

If you look at the diag files, it is pretty clear that the reason the jumps are slower early on that the wind is very optically thick, and that later on it gets thinner as the temperature evolves,  the next few lines were obtained by simply grepping out a line of one of the diag files:


````
No. of photons which have scattered n times.     The max number of scatters seen was 357
No. of photons which have scattered n times.     The max number of scatters seen was 512
No. of photons which have scattered n times.     The max number of scatters seen was 152
No. of photons which have scattered n times.     The max number of scatters seen was 78
No. of photons which have scattered n times.     The max number of scatters seen was 59
No. of photons which have scattered n times.     The max number of scatters seen was 36
No. of photons which have scattered n times.     The max number of scatters seen was 31
No. of photons which have scattered n times.     The max number of scatters seen was 25
No. of photons which have scattered n times.     The max number of scatters seen was 27
No. of photons which have scattered n times.     The max number of scatters seen was 27
No. of photons which have scattered n times.     The max number of scatters seen was 26
No. of photons which have scattered n times.     The max number of scatters seen was 27
No. of photons which have scattered n times.     The max number of scatters seen was 29
No. of photons which have scattered n times.     The max number of scatters seen was 27
No. of photons which have scattered n times.     The max number of scatters seen was 27
No. of photons which have scattered n times.     The max number of scatters seen was 21
No. of photons which have scattered n times.     The max number of scatters seen was 33
No. of photons which have scattered n times.     The max number of scatters seen was 26
No. of photons which have scattered n times.     The max number of scatters seen was 31
No. of photons which have scattered n times.     The max number of scatters seen was 30
No. of photons which have scattered n times.     The max number of scatters seen was 25
No. of photons which have scattered n times.     The max number of scatters seen was 32
No. of photons which have scattered n times.     The max number of scatters seen was 25
No. of photons which have scattered n times.     The max number of scatters seen was 29
No. of photons which have scattered n times.     The max number of scatters seen was 25
No. of photons which have scattered n times.     The max number of scatters seen was 25
No. of photons which have scattered n times.     The max number of scatters seen was 28
No. of photons which have scattered n times.     The max number of scatters seen was 26
No. of photons which have scattered n times.     The max number of scatters seen was 28
No. of photons which have scattered n times.     The max number of scatters seen was 31
No. of photons which have scattered n times.     The max number of scatters seen was 26
````

It would be interesting to understand why this is.  (Note that photon transport dominates the total time in all cases; the time to calculated the matrices need to use the matrix method is not a factor.)

**Comment by kslong** [comment]
This issue has at least partially been addressed by 83d46174dfc32ec0f0943825edc362d44a51ee28 so I am closing it for now
